Feature/36 : Product Server 숙박업체 재고 조회 Service

shbada opened the pull request:
- 재고 프로세스 코드리뷰 요청
- [예정/#46] Exception 처리 진행 예정으로, 이부분은 제외하고 리뷰 부탁드립니다~!

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- Accommodation domain, infrastructure add
- Booking Server -> Product Server module use setting

Commit:
Merge remote-tracking branch 'origin/develop' into feature/36

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- booking -> accommodation service layer call sample 적용

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- accommodation stock 조회 로직 반영

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- booking -> product call 테스트코드 확인
- 샘플 데이터 (Test h2 DB) 용 .sql 파일 추가
- Product com->org package rename

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- component scan add

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- redis 연동 + Redisson sample file add

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- stock decrease 로직 (product server -> booking server), RedissonLockFacade 파일 생성 및 적용

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- stock decrease 로직 (booking server -> product server)

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- Redisson AOP 적용 및 Service 제외 후 단순 수행 테스트 완료

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- Redisson 비즈니스 로직 테스트 코드 수행 완료

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- Redisson 비즈니스 로직 테스트 코드 수행 완료

BAEKJungHo commented on expedia-server/product-server/src/main/java/org/clonecoder/productserver/common/aop/RedissonCallTransaction.java:
```diff
@@ -0,0 +1,20 @@
+public class RedissonCallTransaction {
+    /**
+     * 부모트랜잭션의 유무와 관계없이 동시성에 대한 처리는 별도의 트랜잭션으로 동작하기 위함
+     * @param joinPoint
```
> 부모트랜잭션의 유무와 관계없이 동시성에 대한 처리는 별도의 트랜잭션으로 동작하기 위함

이유가 있을까요?

shbada replied:
위 별도의 트랜잭션 생성은 Redission 적용 학습을 할때 발견한 내용인데요, 제가 이해한 내용에 따르면 아래와 같습니다

코드를 보시면 try~catch~finally 문에서 finally에 unlock() 메서드를 호출하고있습니다.
해당 메서드의 마지막 부분에서 lock을 해제하는데, 부모 트랜잭션은 메서드 로직 수행이 모두 끝난 후에 실행됩니다. (commit)
이때 만약 A쓰레드, B 쓰레드가 있을때,
A쓰레드가 Redis Rlock을 잡고있다가 unlock()이 호출되어 반환을 했는데 (아직 commit 전) B쓰레드가 Rlock을 얻고 재고 수량을 확인할때 DB에는 -1이 되지 않은 재고값이 조회가 되서 데이터가 맞지 않게될 수 있어서,
재고 로직을 수행하는 실제 Service를 별도의 트랜잭션으로 처리해서 이 트랜잭션을 커밋 후 lock()을 해제하도록 하게하기 위함입니다
(Redis가 unlock() 되는 시점 < commit() 시점이라서  unlock() ~ commit() 사이에 redis lock에 이슈가 발생한다거나 등의 상황으로 이해했습니다~)

BAEKJungHo commented on expedia-server/booking-server/src/main/java/org/clonecoder/bookingserver/infrastructure/BookingStoreImpl.java:
```diff
@@ -29,4 +32,12 @@ public Booking saveBooking(Booking booking) {
+     * @param accommodationRoomId
+     */
+    public void accommodationRoomStockDecrease(Long accommodationRoomId) {
+        accommodationService.stockDecrease(accommodationRoomId);
```
infra 계층에서 service 를 사용하여 호출하는 방향은 지양해야하지 않을까 생각합니다.

bookingStoreImpl 은 data access 를 위한 세부 기술을 구현하는 것인데, 이렇게되면
bookingStoreImpl 코드 구현 규칙이 이상해질 것 같아요.

1. storeImpl 에서 -> service 를 참조
2. storeIMpl 에서 -> database 접근

BAEKJungHo commented on expedia-server/product-server/src/main/java/org/clonecoder/productserver/domain/accommodation/AccommodationService.java:
```diff
@@ -0,0 +1,26 @@
+        AccommodationRoom accommodationRoom = accommodationStore.getAccommodationRoom(accommodationRoomId);
+        log.info("stock() : " + accommodationRoom.getStock());
+
+        accommodationRoom.stockDecrease();
```
> 트랜잭션 전파 레벨 옵션과 롤백에 대해서 자세하게 알지 못해서 ㅜ 질문식으로 남겨놓을게요

[Q-1]
1. 별도의 트랜잭션으로 돌고
2. Dirty Chekcing 이 일어나고
3. stockDecrease 메서드가 완료되어 재고 감소가 성공적으로 일어났는데
4. stockDecrease 를 사용하는 bookingService 쪽에서 응답 직전에 서버 장애가 발생하게 되면 주문은 생성되지 않았는데 재고가 감소하는 일이 발생하지 않을까 생각합니다.

[Q-2] - REQURIRES_NEW 로 별도의 트랜잭션으로 돌렸을때, stockDecrease 메서드에서 예외가 발생하면 부모 트랜잭션에서는 해당 예외를 catch 하여 부모 트랜잭션까지 rollback 이되는 부분일까요?

[Q-3] - 반대로 아래 구조라고 가정

```java
bookingService(..) {
  // 1. 주문 생성
  // 2. 재고 감소
  // 3. 부모 트랜잭션과 함께 성공해야 하는 다른 로직
}
```

이런 구조일 때, 2번까지 성공하고 3번에서 에러가 발생한 경우, 1,3 은 롤백되고 2번은 재고가 감소되지 않을까 생각하는데 맞을까요?

shbada replied:
Q-1. 
네네 
> stockDecrease 를 사용하는 bookingService 쪽에서 응답 직전에 서버 장애가 발생하게 되면 주문은 생성되지 않았는데 재고가 감소하는 일이 발생하지 않을까 생각합니다.

이 상황이 발생하는거 맞습니다, 응답 직전에 서버 장애가 발생한 경우 대응을 처리하는게 쉽지는 않아보이네요,
만약에 재고감소 로직 (새로운 트랜잭션 생성) 이후 어떤 서비스 로직이 있다면 재고감소 로직 위치를 해당 메서드의 마지막에 위치를 시키는게 최선의 방법인거같아요

Q-2.
네네, 말씀하신대로 트랜잭션의 전파는 신규 트랜잭션 생성시 전해지지 않지만, 예외를 던지는 경우 부모 트랜잭션에서 예외 처리를 하게되서 롤백되는게 맞습니다~

Q-3. 
네 맞습니다, 우선 위 Q-1, Q-2 상황으로 제가 에러 처리를 해서 데이터 확인까지 해보았었는데요,
Q-3번의 경우 1,3은 롤백되고 2번은 롤백되지않는 이슈가 맞아서 해결방안으로는 아직까지 2번(신규 트랜잭션)의 경우를 메서드이 마지막에 위치하도록 하거나, 3)번 실패시 2)번의 경우를 롤백하는 로직을 추가하는 방안이 있어야할거같습니다,

BAEKJungHo commented on expedia-server/booking-server/src/main/java/org/clonecoder/bookingserver/domain/booking/BookingService.java:
```diff
@@ -34,7 +36,20 @@ public Booking saveBooking(BookingCommand bookingCommand, List<BookingGuestsComm
+     * @param accommodationRoomId
+     */
+    public void accommodationRoomStockDecrease(final Long accommodationRoomId) {
+        bookingStore.accommodationRoomStockDecrease(accommodationRoomId);
```
service 간의 참조관계를 가지지 않게 하기 위해서 이런 구조(storeImpl 에서 service 를 호출하는 구조)를 가져가신 것 같은데,

1. 재고 감소는 언제 일어나는지?
2. 재고 감소 로직을 꼭 AccommodationService 로 빼서 관리해야하는지

위 경우를 조금 만 더 생각해보면, 조금 더 나은 구조가 될 것 같아요. 

> 제 생각에는 현재 구조에서 불필요한 추상화 비용을 줄일 수 있을 것 같아요.

shbada replied:
네 맞습니다, 멀티 모듈의 경우 Service -> Service 간의 참조를 하지 않고, infrastructure 계층에서 accommodation 모듈의 Service 호출로직을 가져가려고 했었는데, 아래 댓글이랑 같이 참고해서 보니, domain 계층에서 처리할 수 있는지 다시 고민해봐야겠씁니다~
재고 감소 로직을 AccommodationService로 빼지않고 booking에서 Accommodation 엔티티를 접근해서 처리할 수는 있지만 숙박 관련 로직은 Accommodation 모듈 내에서 처리하고싶었거든요, 
만약에 booking에서 직접 엔티티에 접근하려면 엔티티를 core module로 옮기는 리팩토링이 진행되는게 맞는데,
우선 Accommodation 관련 로직들을 Accommodation 모듈 내에 구현하는게 추후 관리시에 더 낫지않을까 생각이들었습니다~!

```
infra 계층에서 service 를 사용하여 호출하는 방향은 지양해야하지 않을까 생각합니다.

bookingStoreImpl 은 data access 를 위한 세부 기술을 구현하는 것인데, 이렇게되면
bookingStoreImpl 코드 구현 규칙이 이상해질 것 같아요.

storeImpl 에서 -> service 를 참조
storeIMpl 에서 -> database 접근
```

BAEKJungHo replied:
상품 관련 로직을 상품쪽으로 관리하는 것은 좋다고 생각합니다.

현재는 AccommodationService 에서 stockDecreae 가 하는 역할이 

- DB 조회
- 엔티티에서 stockDecrease 호출

이게 끝이라서, 사실상 도메인 로직은 Entity 에 있는게 맞다고 생각이 들고, 지금 구현도 그렇게 되어있다고 생각이 들어서 AccommodationService 가 당장은 필요 없지 않을까 생각이 들었어요.
즉, `재고 감소`라는 로직은 DomainService 와 Entity 중에서 Entity 에서 다뤄야하는 도메인 로직이므로 BookingService 에서는 `상품엔티티 = 상품Store.조회` -> 상품엔티티.stockDecrease 하는게 추상화 비용을 줄이는 방안이지 않을까 생각했습니다. 

또한, 재고 감소 로직은, __주문(예약)__ 시에만 발생하므로, 특정 방에 대한(accRoommId) 주문 자체에 RedissonLock 을 걸어버리는 것도 괜찮지 않을까 생각했습니다. 

> 물론 위처럼 하려면, seohae 님이 말한 것처럼 엔티티를 core module 로 빼내긴 해야할 것 같아요

BAEKJungHo replied:
> 또한, 재고 감소 로직은, 주문(예약) 시에만 발생하므로, 특정 방에 대한(accRoommId) 주문 자체에 RedissonLock 을 걸어버리는 것도 괜찮지 않을까 생각했습니다.

이 부분은 추가적으로 seohae 님이 검토 해서 적용할지, 상품 재고 감소 로직에만 락을 걸지 정하긴 해야할 것 같습니다.

제가 생각한 내용:

1. 주문 생성
2. 상품 재고 감소
3. 결제 완료

1 -> 2- > 3 이 모두 완료되어야 상품 재고를 감소하는게 맞다고 생각이 들어서 @RedissonLock 을 BookingService 의 주문생성 메서드에 걸어버리는 것도 괜찮지 않을까 생각해봤습니다.

shbada replied:
정호님 말대로 Entity에 단순 로직이라서, 이 Entity를 직접적으로 호출해도 될거긴한데,
1) DB 조회
이때 roomId로 AccommodationRoom 객체를 조회해오는 로직이 존재하고,
이 조회한 객체로
2) stockDecrease 호출
프로세스로 진행되고있는데, 만약 service 로직을 제외한다면 여기서 BookingService -> AccommodationRoomRepository 로 의존성이 생기게되서 이때 StoreImpl에 추가해야하지 않을까 생각이 들었습니다~!

1)번의 로직이 수행되어야할때 AccommodationRepository에 대한 의존성이 생기게되는데, 이때를 고려해서 Accommodation 모듈쪽에 추상화한 메서드를 추가해서 이 클래스를 보도록 했습니다!

BAEKJungHo replied:
> AccommodationRoomRepository 로 의존성이 생기게되서 이때 StoreImpl에 추가해야하지 않을까 생각이 들었습니다~!

DomainService 에서는 Store 를 의존하고 있어도 되기 때문에 

```kotlin
BookingSerivce {
   private val bookingStore: BookingStore
   private val AccStore: AccStore
}
```

위 구조가 되어도 이상한 점은 없을 것 같아요.

> 만약 service 로직을 제외한다면 여기서 BookingService -> AccommodationRoomRepository 로 의존성이 생기면서

위 처럼 하면 bookingservice -> repository 의 의존성은 없고 store interface 에 대한 의존성만 있을 것 같아요.

BAEKJungHo replied:
만약에, BookingService 에서 Acc의 Store Interface 자체를 의존하는게 싫은 경우에는 리팩토링 하려면

1. service 간의 의존관계가 생기거나
2. bookingfacade 에서 2개의 서비스를 호출하거나 (~~이 경우, 도메인 로직이 application 으로 노출됨~~)

둘 중 하나이지 않을까 생각이 들어요.

shbada replied:
```
1) DB 조회
2) stockDecrease 호출
```
아, 1번을 accStore로 옮기고, 조회해온 객체로 BookingService에서 직접 엔티티의 decrease() 를 호출하도록 하는 프로세스를 말하시는거같은데 

1) BookingService -> AccommodationRoomStore(위 1),2) 로직 모두)
2) BookingService -> AccommodationRoomStore(위 1)) -> BookingService 에서 AccmmodationRoom Entity

제가 말했던 Accommodation 모듈 내에서 처리 + store로 가져갔을때 domain -> infrastructure 의존성 이렇게 조건이 모두 맞는 프로세스는 1번이라고 생각하는데 제가 이해한게 맞을까요?!

BAEKJungHo replied:
```java
public class BookingSerivce {
  private BookingStore bookingStore;
  private AccStore accStore; // AccStoreImpl 에서는 Repsitory 를 DI 하고 호출함

  @RedissonLock // Lock 을 서비스에서 거는게 맞을지, Controller 에서 부터 거는게 맞을지 검토 필요
  public 주문 생성(Dto: request) {
    /* booking 저장 */
    /* booking guests List 저장 */
    /* 예약 저장 */
    /* 예약건에 대한 재고 감소 */
    AccEntity entity = accStore.getByRoomId(reqeust.roomId);
    entity.stockDecrease();
  } 
}
```

> 글을 제가 잘 이해 못해서 코드로 남겨놓아요 ㅎㅎ
>
> 결제는 빼뒀습니다.

shbada replied:
아 네네 ㅎㅎ 코드로 보니까 제가 위에서 언급했던 것중에  2)번을 말씀주신게 맞는데 
우선 2)번의 경우가 더 맞는가 라고 했을때 
> Entity를 core module로 올려야한다.

이게 맞는 상황일때 저 코드 로직이 맞아보입니다! 우선 Entity를 core module로 올려서 모든 모듈에서 공유를 하는게 맞을까? 각 엔티티는 각 책임에 맞는 모듈에서 관리되는게 맞지 않은가?라는 의문이 있긴합니다, 이부분은 멀티 모듈 구조에서 어떤게 더 나은지 한번 고민해보겠습니다 ㅎ
지금 아직은 의문인 점은 
```
/* 예약건에 대한 재고 감소 */
    AccEntity entity = accStore.getByRoomId(reqeust.roomId);
    entity.stockDecrease();
```
위 로직을 AccommodationStoreImpl 에 구현해서 AccEntity에 대한 의존성을 Acc module에만 있게하는게 더 나은게 아닌가가 계속 고민이되네요,,

> BookinStoreImpl -> AccommodationService 의존성은 정호님 말대로 domain layer -> infrastructure layer 의존성이 맞아보여서 이부분은 수정하겠습니다~

BAEKJungHo replied:
> 위 로직을 AccommodationStoreImpl 에 구현해서 AccEntity에 대한 의존성을 Acc module에만 있게하는게 더 나은게 아닌가가 계속 고민이되네요,,

저도 회원/인증하면서 그 고민을 했었는데요. 엔티티를 core 모듈로 빼지 않으면 결국 A(Order) -> B(Product) (화살표 방향이 의존관계) 관계에서 A 가 B 를 사용하려면 Rest call 말고는 답이 없는 것 같아요. (서해님이 하신 것처럼 infra 에서 domain 호출을 하거나)

그러면 AccEntity 를 core 모듈로 뺐을 때 어떤 그림이냐 ?

- 주문 도메인 입장에서는 다음과 같음
  - Order -> Product 
    - Order 에서 Product Entity 를 필요로함 -> core 모듈에서 가져다 사용
- 상품 도메인 입장에서는 다음과 같음
  - Core 모듈에 product entity 가 있기 때문에 그냥 가져다 사용
  - product 에서 order 를 참조하진 않음

결국 product entity 를 core 모듈로 빼더라도 의존 관계방향을 해치진 않는다고 생각해요. 

대신, order entity 를 core 모듈로 빼야하나? <-- 이 부분은 고민이 될 것 같아요. 그렇게되면 product 에서도 order entity 를 호출할 수 있게 되기 때문에

BAEKJungHo replied:
> domain layer -> infrastructure layer 의존성이 맞아보여서 이부분은 수정하겠습니다~

domain 은 infra 를 바라보면 안됩니다. 

infra -> domain 을 바라보는 거여서, 제가 말하고자 했던 것은 __infra 에서 domain 을 바라본다고 domain service 를 호출해도 되는 것이냐?__ 가 핵심이었어요

infra 는 db access 과 같은 세부적인 기술들을 모아두는 곳인데, 세부적인 기술들을 모아두는 곳에서 도메인 로직을 호출한다? -> 과연 맞는 방향일까? 였습니다.

마치 storeiml 에서 entity 를 조회하고나서 storeImpl 에서 entity.stockDecrease 를 호출하는것과 똑같다고 생각됩니다.

shbada replied:
> domain layer -> infrastructure layer 의존성이 맞아보여서 이부분은 수정하겠습니다~
> domain 은 infra 를 바라보면 안됩니다.

아아 맞네요, 
infrastructure layer -> domain Layer
이것때문에 초기 코드가 BookingStoreImpl -> AccommodationService로 구현했었습니다~

> infra 에서 domain 을 바라본다고 domain service 를 호출해도 되는 것이냐?

근데 제가 위에처럼 Store 호출 로직을 Service로 옮겼었는데, 그 이유가 조회/재고감소 도 서비스 로직이라고 생각하고 이를 serviceTest로 가져가기위해서였고, 위 문장처럼 infra -> domain 의존성이기 때문에 service로 옮겨서 bookingStoreImpl -> AccommodationService로 가져갔었습니다 (정확하게 캐치하셨네여 ㅎ)

그래서 기존 로직에서 BookingStoreImpl -> AccommodationService의 의존성이 생겨난거였고, 이 부분은 BookingService -> AccommodationStore 방향으로 가보면 될것같습니다~!

shbada replied:
> 저도 회원/인증하면서 그 고민을 했었는데요. 엔티티를 core 모듈로 빼지 않으면 결국 A(Order) -> B(Product) (화살표 방향이 의존관계) 관계에서 A 가 B 를 사용하려면 Rest call 말고는 답이 없는 것 같아요. (서해님이 하신 것처럼 infra 에서 domain 호출을 하거나)

> 결국 product entity 를 core 모듈로 빼더라도 의존 관계방향을 해치진 않는다고 생각해요.

core 모듈로 빼면 booking -> product domain layer 로 가지 않고, booking이 core의 entity를 곧바로 참조할 수 있는 구조가되는 부분인건 맞는데 core로 모든 엔티티를 뺴는것과 몇 공통적으로 사용되는 엔티티만 빼는것, 이 상황들은 뭐가 나은지 좀더 고민해봐야할거같네요

> 상품 도메인 입장에서는 다음과 같음
> Core 모듈에 product entity 가 있기 때문에 그냥 가져다 사용
> product 에서 order 를 참조하진 않음

Core 모듈에 entity 파일을 공통적으로 사용해도 되느냐, 어떤 이슈가 없느냐를 좀더 생각해봐야겠습니다,ㅎㅎ

BAEKJungHo replied:
> 몇 공통적으로 사용되는 엔티티만 빼는것, 이 상황들은 뭐가 나은지 좀더 고민해봐야할거같네요. Core 모듈에 entity 파일을 공통적으로 사용해도 되느냐, 어떤 이슈가 없느냐를 좀더 생각해봐야겠습니다,ㅎㅎ

좋아요:smile: 저도 아직 명쾌한 해답을 발견하지 못한 상태라 ㅎㅎ 좋은 아이디어있으면 같이 얘기해봐요

shbada replied:
> 저도 회원/인증하면서 그 고민을 했었는데요. 엔티티를 core 모듈로 빼지 않으면 결국 A(Order) -> B(Product) (화살표 방향이 의존관계) 관계에서 A 가 B 를 사용하려면 Rest call 말고는 답이 없는 것 같아요. (서해님이 하신 것처럼 infra 에서 domain 호출을 하거나)

근데 엔티티를 core로 빼더라도 만약에 저같이 조회) 등의 비즈니스 로직이 추가된다고 하면 어차피 모듈간의 호출이 발생할거같은데 그렇다면 해당 모듈의 메서드를 호출하는게 더 편리하지 않을까라는 생각도 들어요,

> 아 예를들어서 product server에 이미 저 1)2)번을 수행하는 메서드가 있을때도 booking server에서 1)번의 로직만 필요하니까 product server에 1)번만 수행하는 메서드를 호출하고, 2)번 entity를 호출하는건 별도로 booking server에서 core entity를 사용하는지,
이 상황이 궁금했어여
그렇다고 해당 로직 메서드가 존재하는 경우에는 이 메서드를 호출하고, 아닌 경우에는 1)번은 product, 2)번은 core entity를 사용하고 등등은 상황에 따라 달라지는 코드 구현이라 의도치않은 전략인거같구
사실 결론만 얘기하자면 고민인거는
1)2)번을 수행하는 메서드를 제공하는게 1),2) 따로 (product server, core entity) 보다 좀더 코드관리에 편하지 않을까라는 생각을 했거든영

BAEKJungHo replied:
> 참고. 저는 아래와 같은 구조로 사용하고 있습니다.  의존관계: auth -> member

- 인증/회원
  - auth-server
    - security 라는 네이밍의 패키지(= 도메인 패키지)가 존재 
  - core
    - domain
      - member
       - ...
    - infra
      - MemberRepository, storeImpl
  - member-server
    - application
    - interfaces

shbada replied:
아, core 모듈에 domain, infra 모두 둬서 repository, entity 모두 접근이 가능하도록 되어있군요,
인증/회원은 의존성이 많다보니 개발하시면서 저 구조가 나을거라는 확신이 생기셨을거같아요 ㅎㅎ
저도 우선 좀더 구현해보다가, 정호님처럼 booking-proudct 간의 관계가 저 core 모듈을 사용하는 방향이 더 나은걸로 생각된다면 그때 한번 진행해보겠습니다-!

Commit:
feature36: Product Server 숙박업체 재고 조회 Service
- BookingStoreImpl -> AccommodationService 의존성 제거
- BookingService -> AccommodationStore 추가
